### ISSUE TITLE
Investigate on Gatling slowness on the client for the RealLifeSimulation

### ISSUE THREAD

**Issue by lucamilanesio** [GerritForge/gatling-sbt-gerrit-test#5]
# Background

It is hard to make a single simulation that would represent two very different usages  of the repository:
1. CI/CD: typical use is to clone a repo from scratch (typically using a reference-repo), after the clone there is a delay (simulating a build time) and report a feedback (Verified +1/-1 with some comments).
2. Developer: does an initial clone (not necessarily part of the simulation, it could be a copy of a local mirror) at the beginning of the simulation representing a new dev starting to to work a project, then make local changes, push new changes, add comments and votes on reviews, submit changes or abandon changes, from time to time fetch the remote branches to update the local workspace.

# Entropy

The two scenarios are played by two different types of actor:
1. CI/CD: it does always the same thing, first clone, then fetch an *existing* open change, run the build and lastly provide feedback. (deterministic behaviour)
2. Developer: it represents a human and therefore it does not do the same thing all the times, even though the % of distributions of the actions can be defined and tuned. (non-deterministic behaviour)

# Shape of the parallelism

The two scenarios may have a completely different shape of parallelism when executed:
1. CI/CD: typically a series of spikes when the changes are discovered and it is going up very quickly because of the builds parallelism.
2. Developer: it more a constant traffic, typically follows the working day, with a rapid ramp up between 9 and 10 AM up to lunchtime and a slow ramp down up to 5 PM.

> **NOTE**: The above assumptions are based on a co-located Team, but when projects have multiple Teams on different time-zones, the above curves may overlap with regards to the Developer's shape.

# Requirements

- *AS* a developer of a Gatling simulation
- *WHEN* a run a prototype of 1. or a prototype of 2.
- *THEN* I have the clear directions of what are the fine-grained actions and their randomness
- *SO THAT* I can code and structure the correct sequence of actions in Gatling
- *AND* I can generate a *truly* real-life situation that matches the percentages and usage expected (82% upload-pack, 11% post comment, 4% receive-pack, 2% submit, 1% abandon)
- *AND* I should know what is the split of the 82% upload-pack between the developer (performing just upload-pack on existing working local repos) and the CI/CD (performing full clone using a reference repo + fetch of individual changes)

# Implementation details

For prototyping a simple bash script for 1. and another for 2. would be _good enough_ to satisfy the requirements. For simulating some entropy a simple %100 of the current time in millis should be more than enough.

The objective of the scripts are:
- They are an executable specification and we can run them against Gerrit and observe the % of calls generated by inspecting Gerrit logs
- They are exact in the way they describe the behaviour to be implemented, but they may not be fully suitable as-is for our tests

# Non-requirements

- We **do not want** to implement the parallelism with the prototype scripts, however, we want to give the input specifications for the next phase (creating Gatling tests for scenario 1. CD/CD and 2. Developer) for allowing the implementation of the real Gatling code.
- We **do not want** to implement the true randomness of the actions, as this is just a prototype
- We **do not want** to track execution times on the client side, metrics or performance, as this is just a prototype
- We **do not want** to generate any report
- We **do not want** flexibility and configurability of the prototype

**Comment by lucamilanesio** [comment]
Kick-off with @syntonyze has been published: https://drive.google.com/file/d/1kqygH4m3b6fr1HDZzI_rEeDTQdYAJIK2/view?usp=sharing

**Comment by avilaplana** [comment]
I have some questions:

Q1:
> CI/CD: typical use is to clone a repo from scratch (typically using a reference-repo), after the clone there is a delay (simulating a build time) and report a feedback (Verified +1/-1 with some comments).

Do we have any idea on how much do we need to delay the simulation of the build?

Q2:
> AND I can generate a truly real-life situation that matches the percentages and usage expected (82% upload-pack, 11% post comment, 4% receive-pack, 2% submit, 1% abandon)

Based on what it is said in the previous section:

> Developer: it more a constant traffic, typically follows the working day, with a rapid ramp up between 9 and 10 AM up to lunchtime and a slow ramp down up to 5 PM.

Do we need to provide an idleness operation where nothings happen? As a developer I am not running all the time git operations or Gerrit actions (submit, create change, +1)

Q3:
> AND I should know what is the split of the 82% upload-pack between the developer (performing just upload-pack on existing working local repos) and the CI/CD (performing full clone using a reference repo + fetch of individual changes)

I dont understand this. Please, can you elaborate?

**Comment by syntonyze** [comment]
> I have some questions:
> 
> Q1:
> 
> > CI/CD: typical use is to clone a repo from scratch (typically using a reference-repo), after the clone there is a delay (simulating a build time) and report a feedback (Verified +1/-1 with some comments).
> 
> Do we have any idea on how much do we need to delay the simulation of the build?

I think we should be able to define this within some configurable range, keeping in mind that we still need to be able to honour the specified percentage at the end of the simulation.

> 
> Q2:
> 
> > AND I can generate a truly real-life situation that matches the percentages and usage expected (82% upload-pack, 11% post comment, 4% receive-pack, 2% submit, 1% abandon)
> 
> Based on what it is said in the previous section:
> 
> > Developer: it more a constant traffic, typically follows the working day, with a rapid ramp up between 9 and 10 AM up to lunchtime and a slow ramp down up to 5 PM.
> 
> Do we need to provide an idleness operation where nothings happen? As a developer I am not running all the time git operations or Gerrit actions (submit, create change, +1)

It's a good point @avilaplana , I think having a configurable *pause* action would help.
We could always set it to `zero` if we want to. 

> 
> ~Q3:~ ~> AND I should know what is the split of the 82% upload-pack between the developer (performing just upload-pack on existing working local repos) and the CI/CD (performing full clone using a reference repo + fetch of individual changes)~
> 
> ~I dont understand this. Please, can you elaborate?~

Not all git-upload-packs are clones.
Clones and fetches are represented as git-upload-pack server side, but client side they map to two completely different operations. Let's elaborate more f2f if you want.

**Comment by syntonyze** [comment]
## High level tasks / actions / decision points

### Distinguish clones from fetches

When ponch came up with the 82% distribution of git-upload-pack, we didn't drill down into classifying these between clones and fetches. In order to model these as two separate actions we'd need to go back to the logs for the CUE project on [gerrithub.io](http://gerrithub.io/) and run some heuristics that allows us to separate clones from fetches (operation time, bytes transferred). If they are under a threshold X and Y, we consider them as fetch, otherwise we consider them as clones.


### Modelling of dependency actions

We define main actions, the operations that the user is executing and that the we choose with the script input probability config (i.e. clones_pct, create_changes_pct, abandon_pct, post_comment_pct, and so on).
However, not all main actions are independent, some of them require a _previous_ action (which we will calle dependency action), that needs to be executed prior to that.
For example, create change (main action) is dependent on clone (dependency action).
The dependency actions however, are *not directly* modelled via the input distribution config.
The dependency actions need to be accounted for, when honouring the distribution provided as input.
Alternatively, we need to make all actions *pure*, so that they have not actions they depend upon.

### Define input format

- We need the distributions for each main action
- We need the pauses between each main action
- We need a setup step where given a repo, or a gerrit query we fetch the changes or the refs that we want our scripts to work on.


### Define output format and logs

We would like to be able to consume the result of the script in a way that we can easily plot graphs and calculate averages, stddev, max and min. We should come up with a good format for this.
Also, we want to be able to investigate issues and troubleshoot problems, which requires the script to log operations on a file.
The output should also include a table comparing the configured percentages vs the actual ones.

**Comment by lucamilanesio** [comment]
> ## High level tasks / actions / decision points
> ### Distinguish clones from fetches
> When ponch came up with the 82% distribution of git-upload-pack, we didn't drill down into classifying these between clones and fetches. In order to model these as two separate actions we'd need to go back to the logs for the CUE project on [gerrithub.io](http://gerrithub.io/) and run some heuristics that allows us to separate clones from fetches (operation time, bytes transferred). If they are under a threshold X and Y, we consider them as fetch, otherwise we consider them as clones.

That isn't part of this card and isn't needed. If the CI/CD does 1 clone and 1 review action, we have our answers already by simple maths.

> ### Modelling of dependency actions
> We define main actions, the operations that the user is executing and that the we choose with the script input probability config (i.e. clones_pct, create_changes_pct, abandon_pct, post_comment_pct, and so on). 

There is no requirements to have the percentages configurable.

> However, not all main actions are independent, some of them require a _previous_ action (which we will calle dependency action), that needs to be executed prior to that. For example, create change (main action) is dependent on clone (dependency action).

That is untrue: a developer would always reuse an existing working directory. Please watch again the kick-off discussion as we already covered this point.

> The dependency actions however, are _not directly_ modelled via the input distribution config. 

There is no configurability in the requirements: this is a prototype.

> The dependency actions need to be accounted for, when honouring the distribution provided as input. Alternatively, we need to make all actions _pure_, so that they have not actions they depend upon.

No, we write a prototype with a sequence of actions to cover the use-case. Also there is no requirement to develop any Gatling scripts yet. We can write the prototype with a "logic sequence of operations" without the need to execute them randomly.

> ### Define input format
> * We need the distributions for each main action
> * We need the pauses between each main action
> * We need a setup step where given a repo, or a gerrit query we fetch the changes or the refs that we want our scripts to work on.

None of them are required to be configurable: they can all be hardcoded in the script for the prototype.

> ### Define output format and logs
> We would like to be able to consume the result of the script in a way that we can easily plot graphs and calculate averages, stddev, max and min. We should come up with a good format for this. Also, we want to be able to investigate issues and troubleshoot problems, which requires the script to log operations on a file. The output should also include a table comparing the configured percentages vs the actual ones.

There is no requirement for all of these either: we already have all the logs we need and metrics from Gerrit. There is nothing to develop as part of this issue.

**PLEASE** watch the kick-off and I'm happy to have a second kick-off if the objective of this ticket isn't clear enough.

**Comment by lucamilanesio** [comment]
@avilaplana @syntonyze see an example prototype of the CI/CD simulation at https://review.gerrithub.io/c/GerritForge/gerritforge-health-service/+/1197005.

We need to do the same with the developer, just code a flow of actions for the developer, one after the other.

**Comment by avilaplana** [comment]
@lucamilanesio  Apologies for my question but I get confused with the terminology. What does action mean? is it a low operation like `git clone` or `git push a new change` or `verify +1`? or is something more higher level like client operation like `create a change` or `clone a repo` or `list changes`? 
To avoid confusion maybe we could add as a requirement the list of actions that we want to use in the developer script.

**Comment by lucamilanesio** [comment]
> @lucamilanesio Apologies for my question but I get confused with the terminology. 

Never apologise for asking questions :-)

> What does action mean? is it a low operation like `git clone` or `git push a new change` or `verify +1`? or is something more higher level like client operation like `create a change` or `clone a repo` or `list changes` (each client operation might require a set of low level operations)?

It is always a _end-user_ operation: creating a change means _"create a file, add to the staging area, create a commit, push to refs/for/master"_ and so forth.

> To avoid confusion maybe we could add as a requirement the list of actions that we want to use in the developer script.

The high-level actions, it will be clearer once I push the samples.

**Comment by lucamilanesio** [comment]
With regards to the split (82% upload-pack, 11% post comment, 4% receive-pack, 2% submit, 1% abandon) we can assume that:
- 82% upload-pack = 6% clones + 76% fetches
- 11% post comments =  6% as verification of the CI job doing clones + 5% developers' comments
- 4% receive-pack = all create changes made by a developer
- 2% submit = all submissions made by a developer
- 1% abandon = all abandon made by a developer

**Comment by lucamilanesio** [comment]
See some example of prototypes at:
- https://review.gerrithub.io/c/GerritForge/gerritforge-health-service/+/1197005 (CI/CD prototype)
- https://review.gerrithub.io/c/GerritForge/gerritforge-health-service/+/1197014 (Developer)

The above are just example of the level of simplicity of the prototypes, which is purely directed in focusing of what we want to simulate, with imperfect randomness, which is perfectly fine at this stage.

Feel free to drop them and re-do them from scratch, but let's focus on what we want to obtain.

**Comment by lucamilanesio** [comment]
@syntonyze @avilaplana today's checkpoint recording is at https://drive.google.com/file/d/1QLfJzLAHr2V5Siw64PfN9GpmlCXk2OZa/view?usp=sharing

**Comment by syntonyze** [comment]
First run executing [1] for **one** hour.

```
export GERRIT_USER=gatling
export GERRIT_URL=http://ghs-gym.gerritforge.com:8080
export GERRIT_AUTH_URL=http://gatling:secret@ghs-gym.gerritforge.com:8080

nohup ./run_cicd_workflows.sh &
nohup ./run_developer_workflows.sh &
```

```
919 git-upload-pack (grep /git-upload-pack httpd_log | grep -v 401 | wc -l)
1219 git-receive-pack (grep /git-receive-pack httpd_log | grep -v 401 | wc -l)
406 submit (grep /submit httpd_log | wc -l)
179 rebase (grep /rebase httpd_log | wc -l)
644 reviews ( grep /review httpd_log | wc -l)
895 changes have been created
```

[1] https://ghs.gerrithub.io/c/GerritForge/gerritforge-health-service/+/1197169

**Comment by lucamilanesio** [comment]
What's missing is:
- CI scenario that does the polling (4x times more than the CI build)
- Developer abandoning one change

Once we have the two above we can re-run [1].

[1] https://ghs.gerrithub.io/c/GerritForge/gerritforge-health-service/+/1197169

**Comment by avilaplana** [comment]
I have added the abandon query:
```
grep /git-upload-pack httpd_log | grep -v 401 | wc -l
grep /git-receive-pack httpd_log | grep -v 401 | wc -l
grep /submit httpd_log | wc -l
grep /rebase httpd_log | wc -l
grep /review httpd_log | wc -l
grep /abandon httpd_log | wc -l
```

**Comment by avilaplana** [comment]
We started running the ci/cd workflows (pid 606007) in parallel with the developer workflows (pid 606330) at 1:30 pm BST and the number of operations are:

```
git-upload-pack count: 1780
git-receive-pack count: 2412
submit count: 750
rebase count: 375
review count: 1349
abandon count: 376
```

Just in case we have concerns of % of file system used:

`ghs-gatling` machine:
```
/dev/nvme0n1p4  874G  158G  672G  20% /
```

`ghs-gym` machine:
```
/dev/nvme0n1p4  874G  173G  658G  21% /
```
